### PR TITLE
Masked modeling

### PR DESCRIPTION
**Pull request by jmeld**
1. in config/training_args/base.yaml changed "evaluation_steps" to "eval_steps" to be comply with warning about soon to be deprecated key word arguments.
2. created src/model/custom_electra which holds src/model/custom_electra/electra.py and src/model/custom_electra/__init__.py. These allow for passing our custom electra model training.

### CODE REVIEW

File: src/model/custom_electra/electra.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        **kwargs,
+    ):  
+        print("===== kwargs =====")
+        for key, value in kwargs.items():
+            print(f"{key}: {value}")
+        print(f'pretraineds_model_name_or_path: {pretrained_model_name_or_path}')
+        #super().__init__(pretrained_model_name_or_path)
+        super().__init__(**kwargs)
+        self.vocab_size= vocab_size
+        self.generator_config= generator_model
```
Comment (shanest) [inline src/model/custom_electra/electra.py:31]: Should this be `self.generator_model` instead of `.generator_config` (and same for the discriminator below?

Comment (shanest) [reply src/model/custom_electra/electra.py:31]: Actually, I think the variables in the `kwargs` should just be named `_config` instead of the `_model` as they are now

File: src/model/custom_electra/electra.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+from transformers.models.electra.modeling_electra import ElectraForPreTrainingOutput
+
+
+
+class CustomElectraConfig(PretrainedConfig):
+    model_type= "custom_electra"
+
+    def __init__(
+        self,
+        generator_model: transformers.DebertaV2Config,
```
Comment (shanest) [inline src/model/custom_electra/electra.py:16]: (1) rename to `generator_config` (and same for `discriminator`) below.  (2) Just to be fully clear, we don't want to just replace `ElectraX` with `DebertaX`, but rather the most general possible thing that consumes both.  For configs, I honestly think that's just `PretrainedConfig` here

File: src/model/custom_electra/electra.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        super().__init__(config)
+
+        self.lambda_loss = self.config.discriminator_lambda_loss
+        self.gumbel_dist = torch.distributions.gumbel.Gumbel(0.,1.)
+
+        # can change this to pass the configs,
+        # or pass the generator/discriminators themselves
+        #base_gen_config = ElectraConfig.from_pretrained('google/electra-base-generator')
+        base_gen_config= AutoConfig.from_pretrained(config.generator_config)
+        self.generator: ElectraForMaskedLM = AutoModelForMaskedLM.from_config(base_gen_config)
```
Comment (shanest) [inline src/model/custom_electra/electra.py:59]: Type annotation here should just be either `: PretrainedModel` or `AutoModelForMaskedLM`

File: src/model/custom_electra/electra.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        #base_disc_config = ElectraConfig.from_pretrained('google/electra-base-discriminator')
+        base_disc_config = AutoConfig.from_pretrained(config.discriminator_config)
+        self.discriminator: ElectraForPreTraining = AutoModelForMaskedLM.from_config(base_disc_config)
```
Comment (shanest) [inline src/model/custom_electra/electra.py:63]: For the discriminator, we don't want `MaskedLM`, since the task here is replaced token detection.  I'd say there are two options: (i) use `AutoModelForTokenClassification` instead of `AutoModelForMaskedLM`, or (ii) just use `AutoModel`, and add the binary token classification head to this class in `__init__` and `forward`.  Option (i) is probably the smoothest

File: src/model/custom_electra/electra.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        # we can modify src/train_tokenizer.py to take different tokenizer classes
+        # and train it there, pass it here
+        #self.tokenizer = tokenizer if tokenizer else PretrainedTokenizerFast.from_pretrained("google/electra-base-generator")
+
+    
+    def forward(self, 
+                input_ids: Optional[torch.Tensor] = None, 
+                attention_mask: Optional[torch.Tensor] = None, 
+                token_type_ids: Optional[torch.Tensor] = None, 
+                labels: Optional[torch.Tensor] = None) -> ElectraForPreTrainingOutput:
```
Comment (shanest) [inline src/model/custom_electra/electra.py:74]: As per above: return type could be `TokenClassifierOutput`: https://huggingface.co/docs/transformers/v4.44.2/en/main_classes/output#transformers.modeling_outputs.TokenClassifierOutput

File: src/model/custom_electra/electra.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            discriminator_input = input_ids.clone()
+            discriminator_input[labels] = generator_tokens
+        
+        # pass to discriminator
+        discriminator_output = self.discriminator(discriminator_input, attention_mask, token_type_ids, labels=labels)
+
+        # loss
+        loss = generator_output.loss + discriminator_output.loss * self.lambda_loss #might have to rewrite as loss= generator_output.loss.loss + discriminator_output.loss.loss * self.lambda_loss or it might be that discriminator_output.loss.loss is the only one that needs to be .loss twice
+        # must return some output.loss in order to work with trainer
+        return ElectraForPreTrainingOutput(
```
Comment (shanest) [inline src/model/custom_electra/electra.py:102]: As above, I think this can just be a `TokenClassifierOutput` for max generality https://huggingface.co/docs/transformers/v4.44.2/en/main_classes/output#transformers.modeling_outputs.TokenClassifierOutput

**Review by shanest — CHANGES_REQUESTED** [review]
Looking great!  Just a few little changes here and there to get the right level of generality still

**Comment by jmeld** [comment]
Thank you for the prompt and informative suggestions.

**Review by shanest — CHANGES_REQUESTED** [review]
Looks great!  Can you run some quick tests of it first, e.g. passing some inputs through?  I think there should be some errors around `TokenClassifierOutput` and the like, since those aren't imported up top, so doing that kind of running as a test should turn up all of them.  (Running a linter like `mypy` or `pyright` will also turn these up)

**Comment by jmeld** [comment]
Thank you! I implemented the suggestions, and the code runs. I'm not sure why, but pyright highlights some code as breaking type constraints, but I think it might be mistaken.

**Comment by shanest** [comment]
Hi Jon: just circling back to what we discussed in the meeting the other day.  Can you and/or Diana work on integrating the custom Electra/RTD classes in BabyLM repo and getting things working there?  I'd like to merge this PR into main, but without the custom model classes :)

**Comment by jmeld** [comment]
Yes. I have

1. Removed the src/model/custom_electra/__init__.py and src/model/custom_electra/electra.py files from this lm-training repo to clean it up.
2. In the babylm branch da/electra I added a new directory babylm and made it into a package by adding babylm/__init__.py, and added a sub-package named model with babylm/model/electra.py and babylm/model/__init__.py

**Comment by kharaldsson** [comment]
> Yes. I have
> 
> 1. Removed the src/model/custom_electra/**init**.py and src/model/custom_electra/electra.py files from this lm-training repo to clean it up.
> 2. Made the changes you suggest here to electra.py in the babylm git repo's da/electra branch. So this pull request should clean up the lm-training repo
> 3. Added some new tokenizer training configs for training masked language modeling but adding a MASK, CLS, and SEP token instead of just a MASK token (as the previous masked training model does).

@shanest are there any outstanding comments on this PR? I'm assuming it's a blocker for bablym?

**Comment by shanest** [comment]
Thanks for digging this back up!  And thanks Jon for cleaning things up. Yet, looks good to me now, will merge :).  We'll need to go back to the babylm repo and check out the modeling PR over there for the Next Token Prediction stuff

**Review by shanest — APPROVED** [review]
